mm2src/coins/utxo/rpc_clients.rs

```diff
@@ -2388,32 +2546,445 @@ impl UtxoRpcClientOps for ElectrumClient {
+        )
+        .await?;
+
+        let suspend_timeout_sec = Self::get_suspend_timeout(&guard, &address).await?;
+        Self::duplicate_suspend_timeout(&mut guard, &address).await?;
+        drop(guard);
+
+        self.0.abortable_system.weak_spawner().spawn(async move {
+            debug!("Suspend server: {}, for: {} seconds", address, suspend_timeout_sec);
+            tokio::time::sleep(Duration::from_secs(suspend_timeout_sec)).await;
+            self.restore_server(address).await;
+        });
+
+        debug!("Suspend future spawned");
+        Ok(())
+    }
+
+    async fn cancel_connection(
+        state: &mut MutexGuard<'_, ConnMngState>,
+        address: &str,
+        abortable_system: AbortableQueue,
+    ) -> Result<(), String> {
+        debug!("Cancel connection: {}", address);
+        let mut conn_ctx = state.conn_ctxs.get_mut(address).expect("TODO");
+        conn_ctx.abortable_system.abort_all().map_err(|err| {
+            ERRL!(
+                "Failed to abort on electrum connection related spawner: {}, error: {:?}",
+                address,
+                err
+            )
+        })?;
+        conn_ctx.connection.take();
+        conn_ctx.abortable_system = abortable_system;
+        Ok(())
+    }
+
+    async fn get_suspend_timeout(state: &MutexGuard<'_, ConnMngState>, address: &str) -> Result<u64, String> {
+        state
+            .conn_ctxs
+            .get(address)
+            .map(|ctx| ctx.suspend_timeout_sec)
+            .ok_or_else(|| {
+                ERRL!(
+                    "Failed to get suspend_timeout for address: {}, use default value: {}",
+                    SUSPEND_TIMEOUT_INIT_SEC,
+                    address
+                )
+            })
+    }
+
+    async fn duplicate_suspend_timeout(state: &mut MutexGuard<'_, ConnMngState>, address: &str) -> Result<(), String> {
+        let suspend_timeout = &mut state
+            .conn_ctxs
+            .get_mut(address)
+            .ok_or_else(|| ERRL!("Failed to duplicate suspend_timeout for address: {}, no entry", address))?
+            .suspend_timeout_sec;
+        let new_timeout = *suspend_timeout * 2;
+        debug!(
+            "Duplicate suspend timeout for address: {} from: {} to: {}",
+            address, suspend_timeout, new_timeout
+        );
+        *suspend_timeout = new_timeout;
+        Ok(())
+    }
+
+    async fn reset_suspend_timeout(state: &mut MutexGuard<'_, ConnMngState>, address: &str) -> Result<(), String> {
+        let suspend_timeout = &mut state
+            .conn_ctxs
+            .get_mut(address)
+            .ok_or_else(|| ERRL!("Failed to duplicate suspend_timeout for address: {}, no entry", address))?
+            .suspend_timeout_sec;
+        debug!(
+            "Reset supsend timeout for address: {} from: {} to the initial value: {}",
+            address, suspend_timeout, SUSPEND_TIMEOUT_INIT_SEC
+        );
+        *suspend_timeout = SUSPEND_TIMEOUT_INIT_SEC;
+        Ok(())
```

Comment:
Maybe these 2 functions can be made into one, it can take a parameter `action: TimeoutAction` to know whether to reset or duplicate the timeout.

Reply:
It may not be necessary to introduce a new enumeration; the method names are sufficiently mnemonic and meaningful. Therefore, I took the liberty of suggesting [another approach](https://github.com/KomodoPlatform/komodo-defi-framework/pull/1966/commits/d98521f35a4ee4d682d49b248d3d3ff04929af4f)